### PR TITLE
`dagger do` action options flags

### PR DESCRIPTION
**Pull request by jlongtine**
Had to recreate #1902 because of private -> public transition

Signed-off-by: Joel Longtine <joel@dagger.io>

### CODE REVIEW

**Review by TomChv — COMMENTED** [review]
Is it ready to be reviewed? Last one was a draft and I still see unfinished things in the code

**Comment by jlongtine** [comment]
> Is it ready to be reviewed? Last one was a draft and I still see unfinished things in the code

@TomChv Not yet. Still need a run through of the DX with @shykes. Then I'll clean up all of the code and submit for review. I'll change it to draft. Forgot to do that when I created a new PR.

**Comment by helderco** [comment]
Hey @jlongtine, I need this for #1351. What do you need to move forward?

**Comment by helderco** [comment]
Hey @jlongtine, I need this for #1351. What do you need to move forward?

**Comment by shykes** [comment]
> Hey @jlongtine, I need this for #1351. What do you need to move forward?

We went through DX together, it all looks good, no issue on my end (unless something changed in the last couple days).

**Comment by jlongtine** [comment]
@helderco @shykes I had set this aside to focus on #1711, then got caught up in the gears on that one. I was planning on trying to get unblocked there, then moving back onto this. Happy to re-arrange priorities, if need be.

**Comment by jlongtine** [comment]
@helderco After a chat with @aluzzardi, I'm going to focus on this and get it across the line.

**Comment by shykes** [comment]
We're rooting for you @jlongtine :)

**Comment by jlongtine** [comment]
Because I'm actually checking to see if the request path is an action (i.e. has a `cue/flow` Task in it), this causes a behavioral change (hence the change in `bats` tests). I'm not currently sure how to maintain the current `main` behavior and give nice help messages when someone asks for an action that we don't have in our data structures. This is related to #1789.

**Comment by helderco** [comment]
I see. I think that's ok to change. Instead of allowing to `dagger do` a scalar, only allow running actions (i.e., anything that includes a core action no matter the depth).

**Comment by helderco** [comment]
In the case of the `push.cue` test:

```diff
		pullOutputFile: core.#ReadFile & {
			input: pull.output
			path:  "/output.txt"
		pullOutputFile: core.#ReadFile & {
			input: pull.output
			path:  "/output.txt"
+		} & {
+			// Check output file in the pulled image
+			contents: randomString.contents
		}
-
-		// Check output file in the pulled image
-		pullContent: string & pullOutputFile.contents & randomString.contents
```

**Comment by jlongtine** [comment]
@helderco That test is still valid, actually. CUE will fill those values and check that they are the same.

**Comment by helderco** [comment]
> @helderco That test is still valid, actually. CUE will fill those values and check that they are the same.

I know, but you seem to have had to change bats to target `pullOutputFile` instead of `pullContent`. Wasn't that because you can't target `pullContent` with this PR?

**Comment by jlongtine** [comment]
> I know, but you seem to have had to change bats to target `pullOutputFile` instead of `pullContent`. Wasn't that because you can't target `pullContent` with this PR?

Yes, that's correct.

**Comment by shykes** [comment]
Just flagging a potential issue: I've actually ended up writing pure cue actions in at least once case, when I only process inputs without the need for core actions. This is how I hit #1789. Is it possible to ensure that regular structs are taken into account by `dagger do`, and listed in the help message?

I think in the future this problem might go away with #1598, if it defines an explicit `dagger.#Action` that action developers must extend. Then we would no longer have to guess: only instances of that definition would count as an action.

**Comment by helderco** [comment]
> Is it possible to ensure that regular structs are taken into account by `dagger do`, and listed in the help message?

It would be great, problem is, cue/flow doesn't run those.

**Comment by shykes** [comment]
> > Is it possible to ensure that regular structs are taken into account by `dagger do`, and listed in the help message?
> 
> 
> 
> It would be great, problem is, cue/flow doesn't run those.

I don't think it needs to, as long as it runs the client actions wired to it (both inputs and outputs).

**Comment by helderco** [comment]
Yes, true. There's still the case of targetting a scalar or something that doesn't reference core actions eventually, but maybe that's out of scope for this PR.

What should happen with `pullContent` [above](https://github.com/dagger/dagger/pull/2033#issuecomment-1098536702)? I'd say no inputs/outputs in that case. You can wrap a struct over it to enable them.

File: tests/plan/do/do2.cue [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package main
```
Comment (aluzzardi) [inline tests/plan/do/do2.cue:1]: nit: more descriptive test file name

**Review by aluzzardi — APPROVED** [review]
Looks good! Left a minor comment

**Review by helderco — COMMENTED** [review]
Just realized I may have not submitted these comments. They were pending, from before merging.

**Pull request merged**